compiler/transforms/set_memory_space.py

```diff
@@ -0,0 +1,226 @@
+            for use in operand.uses:
+                if (
+                    isinstance(use.operation, memref.MemorySpaceCast)
+                    and isinstance(use.operation.dest.type, memref.MemRefType)
+                    and use.operation.dest.type.memory_space
+                    == builtin.IntegerAttr(1, builtin.i32)
+                ):
+                    cast_op = use.operation
+                    break
+            # If cast op not found, create and insert new one
+            if cast_op is None:
+                cast_op = memref.MemorySpaceCast.from_type_and_target_space(
+                    operand, operand.type, builtin.IntegerAttr(1, builtin.i32)
+                )
+                rewriter.insert_op_before_matched_op(cast_op)
+
+            return cast_op
+
+        # cast all inputs and outputs to correct memory space
+        new_inputs = [
+            inp if get_cast_op(inp) is None else get_cast_op(inp).dest
+            for inp in op.inputs
+        ]
+        new_outputs = [
+            out if get_cast_op(out) is None else get_cast_op(out).dest
+            for out in op.outputs
+        ]
+
+        # new linalg op with new inputs & outputs
+        linalg_op = linalg.Generic(
+            new_inputs,
+            new_outputs,
+            rewriter.move_region_contents_to_new_regions(op.regions[0]),
+            op.indexing_maps,
+            op.iterator_types,
+            op.doc,
+            op.library_call,
+        )
+
+        # replace op
+        rewriter.replace_matched_op(linalg_op)
+
+
+class RealizeMemorySpaceCasts(RewritePattern):
+    """Realize the inserted memory space casts. In the snitch
+    cluster case, the different clusters can only access their own
+    local TCDM L1 memory, so memory space casts are handled by local
+    allocations and memref copies at the correct time.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.MemorySpaceCast, rewriter: PatternRewriter):
+        # check for invalid memory space cast,
+        # no results => unused, removed by cse
+        if len(op.results) < 1:
+            return
+
+        # remove cast, unused op: will be removed by cse
+        if len(op.results[0].uses) == 0:
+            return
+
+        # check if result is memreftype to make pyright happy
+        if not isinstance(op.results[0].type, memref.MemRefType):
+            return
+
+        # replace cast with allocation
+        alloc_op = memref.Alloc.get(
+            op.results[0].type.get_element_type(),
+            64,  # default 64 alignment (necessary ?)
+            op.results[0].type.get_shape(),
+            op.results[0].type.layout,
+            op.results[0].type.memory_space,
+        )
+
+        # Insert copy ops if newly allocated memref is used as
+        # input or output, list to visit all uses of allocated memrefs:
+        uses = [x.operation for x in op.results[0].uses]
+
+        # insert copy to for first use as input
```

Comment:
```suggestion
        # insert "copy to" for first use as input
```
Reads a bit weird otherwise